### ISSUE TITLE
Propose to switch from AStyle to clang-format

### ISSUE THREAD

**Issue by Thanatomanic** [RawTherapee/RawTherapee#6125]
I just committed a `.clang-format` file to facilitate code formatting using the clang toolchain https://github.com/Beep6581/RawTherapee/commit/310239dfbd659fb0d0612e279d6602d1d7dcbc30.
This works particularly well when working in VSCode, for which you only need to install a single extension, whereas getting AStyle to work requires a few annoying manual steps. Even without using an IDE, `clang-format` is readily available in MSYS2 on Windows and easy to install on Linux. How easy this is on macOS I don't know.

**In general I would like to propose to switch from AStyle to `clang-format`.** Development on AStyle seems to have halted (https://sourceforge.net/p/astyle/code/HEAD/tree/trunk/) contrary to clang. At the very least this means that formatting of new language functions (C++20) will be better supported by clang-format than by AStyle.

Does anyone have an opinion on this? @Beep6581 @Floessie @heckflosse @Desmis @Lawrence37 @rom9 

P.S. Regardless of the outcome, I will try to clear up the messy code formatting we currently have, which is all over the place and has not been run through AStyle for a while I presume...

**Comment by Beep6581** [comment]
I don't know this tool, but I support the arguments for using it.

In Sabayon Linux it comes part of sys-devel/clang

Running any code reformatter will break simple merges. Let's wait till after the release.

**Comment by Pandagrapher** [comment]
@Thanatomanic 
> How easy this is on macOS I don't know.

To answer, this tool is available with most popular package manager:
- Homebrew --> Package `clang-format`
- Macports --> Part of package `clang`

Pierre

**Comment by Desmis** [comment]
But simple question....how to use ?

Jacques

**Comment by Thanatomanic** [comment]
@Desmis It depends on how you do your editing. https://clang.llvm.org/docs/ClangFormat.html has the documentation. It can be integrated into vim, emacs, VSCode and a few more. If you would want to test it on a single file, you would do something like:
```
cd ~/RawTherapee/rtgui
clang-format --style=file inputfile.cc > outputfile.cc
```
to produce a new output file, or 
```
clang-format --style=file -i inputfile.cc 
```
to immediately overwrite the inputfile with the changes.

`clang-format` will look for the nearest `.clang-format` file that specifies the styling options and use that.

**Comment by Desmis** [comment]
@Thanatomanic 

 Excuse my ignorance, but do we run this from mingw64, or batch windows, or something else...my clang folders (clang32 clang64) are empty in Msys64 (only subfolders empty..)

jacques

**Comment by heckflosse** [comment]
@Desmis Please test this in msys2 console

```
clang-format --help
```

If that works, you're ready to use it.

**Comment by Desmis** [comment]
@heckflosse 
When I run Msys2 then clang-format --help

I have that
jacques@pc-bureau MSYS ~
$ clang-format --help
-bash: clang-format: command not found

jacques@pc-bureau MSYS ~
$

**Comment by heckflosse** [comment]
@Desmis please try

```
pacman -S mingw-w64-x86_64-clang-tools-extra
```

**Comment by Desmis** [comment]
@heckflosse 
Ingo
After installation...
then I launch  - Mingw64
then
$ clang-format --help
Now it seems to work :)

Thank you

But, just a remark...it's not obvious :)

Jacques

**Comment by heckflosse** [comment]
> it's not obvious

If we want to use it, we need to document it...

**Comment by Desmis** [comment]
I'm talking about the installation, the _a priori_ "everybody knows clang" and not the use which will seem obvious when you're going to use it.

:)

jacques

**Comment by heckflosse** [comment]
> I'm talking about the installation

That's what I meant: Document the installation in rawpedia is needed if we want to use it

**Comment by Desmis** [comment]
ok, yes... :)

jacques

**Comment by rom9** [comment]
@Thanatomanic i've tried on my Fedora Linux (where the tool is provided by the `clang-tools-extra` package).
Il seems that the `SortIncludes: Never` syntax is only supported in clang-format version 12, (while for example Fedora 33 stable is still on v11). This is the error i get:
```
YAML:13:15: error: invalid boolean
SortIncludes: Never
              ^~~~~
```
Changing to `SortIncludes: false` fixes the problem, but i'm afraid it would fail in v12 ; by looking at the documentation, this option doesn't seem to be backwards-compatible.
Anyway, this was just a note in case someone else runs into this problem; if we decide to go for v12, that's :+1:  for me ;-)

alberto

**Comment by Lawrence37** [comment]
I am OK with using clang-format. However, with the current options, it discards manual line breaks in some expressions.

Before:
```c++
bool LCurveParams::operator ==(const LCurveParams& other) const
{
    return
        enabled == other.enabled
        && lcurve == other.lcurve
        && acurve == other.acurve
        && bcurve == other.bcurve
        && cccurve == other.cccurve
        && chcurve == other.chcurve
        && lhcurve == other.lhcurve
        && hhcurve == other.hhcurve
        && lccurve == other.lccurve
        && clcurve == other.clcurve
        && brightness == other.brightness
        && contrast == other.contrast
        && chromaticity == other.chromaticity
        && avoidcolorshift == other.avoidcolorshift
        && rstprotection == other.rstprotection
        && lcredsk == other.lcredsk;
}
```

After:
```c++
bool LCurveParams::operator==(const LCurveParams &other) const
{
    return enabled == other.enabled && lcurve == other.lcurve && acurve == other.acurve && bcurve == other.bcurve && cccurve == other.cccurve && chcurve == other.chcurve && lhcurve == other.lhcurve && hhcurve == other.hhcurve && lccurve == other.lccurve && clcurve == other.clcurve && brightness == other.brightness && contrast == other.contrast && chromaticity == other.chromaticity && avoidcolorshift == other.avoidcolorshift && rstprotection == other.rstprotection && lcredsk == other.lcredsk;
}
```

**Comment by Floessie** [comment]
Great. Saves many LoC. :grimacing: 

Good catch, @Lawrence37. :+1:

**Comment by Floessie** [comment]
```c++
bool LCurveParams::operator==(const LCurveParams &other) const
```

There's also a reason for a space between `operator` and operation, if one considers conversion operators. `operatorbool()` is hard to parse for humans. :wink:

**Comment by ghost** [comment]
FYI clang-format is natively available on Windows as either part of the LLVM / Clang download package, or as part of the Visual Studio free edition install in \VC\Tools\Llvm\x64\bin\clang-format.exe under the MSVC install directory if the LLVM tools are installed as part of C++ dev tools.
